leasing/management/commands/import_level_check_index.py

```diff
@@ -0,0 +1,299 @@
+
+    columns: list[ColumnItem]
+    comments: list[CommentItem]
+    data: list[DataPoint]
+    metadata: list[MetadataItem]
+
+
+class ResponseDataError(Exception):
+    """When the data from API response might cause the program to malfunction."""
+
+    pass
+
+
+class Command(BaseCommand):
+    help = "Import indexes related to Level Check (Tasotarkistus) from Tilastokeskus database."
+
+    def handle(self, *args, **options):
+        for index_input in INDEXES_TO_IMPORT:
+            self.stdout.write(f'Index: "{index_input["name"]}"')
+
+            index_data = _get_index_data(index_input["url"], index_input["code"])
+            columns = index_data.get("columns", [])
+
+            # Find the order of the necessary columns to avoid future failures
+            # if columns are added, removed, or moved.
+            try:
+                quarter_key_pos = _find_key_position(columns, "Vuosineljännes")
+                region_key_pos = _find_key_position(columns, "Alue")
+                index_column_pos = _find_column_position(columns, index_input["code"])
+                number_value_pos = _find_value_position(columns, index_input["code"])
+            except ResponseDataError as e:
+                self.stderr.write(f"Error: {e}. Skipping this index import")
+                continue
+
+            index_column_details = columns[index_column_pos]
+
+            metadata = index_data.get("metadata", [])
+            metadata = next(iter(metadata), {})
+
+            index, created = (
+                OldDwellingsInHousingCompaniesPriceIndex.objects.update_or_create(
+                    code=index_column_details.get("code", None),
+                    url=index_input["url"],
+                    defaults={
+                        "name": index_column_details.get("text", ""),
+                        "comment": index_column_details.get("comment", ""),
+                        "source": metadata.get("source", ""),
+                        "source_table_updated": _get_update_date(metadata),
+                        "source_table_label": metadata.get("label", ""),
+                    },
+                )
+            )
+            self.stdout.write(
+                "Added new index." if created else "Updated existing index."
+            )
+
+            data_points = index_data.get("data", [])
+            self.stdout.write(f"{len(data_points)} data points found.")
+
+            comments = index_data.get("comments", [])
+
+            for dp in data_points:
+                year, quarter = _extract_numbers_from_quarter_name(
+                    dp["key"][quarter_key_pos]
+                )
+                number = _cast_index_number_to_float_or_none(
+                    dp["values"][number_value_pos]
+                )
+                region = dp["key"][region_key_pos]
+                comment = _find_comment_for_value(dp, comments, columns)
+                IndexNumber.objects.update_or_create(
```

Comment:
Do we need any safe guards for obvious errors in the data that would cancel updating any values. For instance if the number is None and we have a previous number, do we want to overwrite the previous number? Perhaps the Model does not allow nulls, but it is not visible here without checking

Reply:
This is something I was wondering too. This could be solved by checking the previous value before updating, but this adds more database calls for each number.

Or by supplying an illegal value as the default, so that the model rejects the update if we don't want it. This would require more error handling to catch this expected exception, so the import run doesn't stop.

This model allows for nulls, so the numbers will be overwritten if the API response would replace a previously existing number value with a period `.`. But I guess there might be some legitimate reason for Tilastokeskus to do this change, and our database should probably reflect the state that is provided by API. I don't see this as a big problem, because if the overwritten number would later receive a proper value later in the API source table, then our numbers would be overwritten again with the proper value.

Reply:
I didn't change the null/empty value handling yet, we can discuss it more if needed.